etl/extractor/platform_impl.go

```diff
@@ -33,220 +30,160 @@ type PlatformExtractor struct {
 	client           exporter.RecordExporterClient
 	request          *exporter.GetRecords
 	mainJetDropsChan chan *types.PlatformJetDrops
+	cancel           context.CancelFunc
+
+	batchSize                                 uint32
+	continuousPulseRetrievingHalfPulseSeconds uint32
 }
 
-func NewPlatformExtractor(batchSize uint32, pulseExtractor interfaces.PulseExtractor, exporterClient exporter.RecordExporterClient) *PlatformExtractor {
+func NewPlatformExtractor(batchSize uint32, continuousPulseRetrievingHalfPulseSeconds uint32,
+	pulseExtractor interfaces.PulseExtractor, exporterClient exporter.RecordExporterClient) *PlatformExtractor {
 	request := &exporter.GetRecords{Count: batchSize}
 	return &PlatformExtractor{
-		stopSignal:       make(chan bool, 1),
 		startStopMutex:   &sync.Mutex{},
 		client:           exporterClient,
 		request:          request,
 		mainJetDropsChan: make(chan *types.PlatformJetDrops),
 
 		pulseExtractor:       pulseExtractor,
 		pulseExtractAttempts: 50,
+		batchSize:            batchSize,
+		continuousPulseRetrievingHalfPulseSeconds: continuousPulseRetrievingHalfPulseSeconds,
 	}
 }
 
-func (m *PlatformExtractor) GetJetDrops(ctx context.Context) <-chan *types.PlatformJetDrops {
-	return m.mainJetDropsChan
+func (e *PlatformExtractor) GetJetDrops(ctx context.Context) <-chan *types.PlatformJetDrops {
+	return e.mainJetDropsChan
 }
 
-func (m *PlatformExtractor) LoadJetDrops(ctx context.Context, fromPulseNumber int64, toPulseNumber int64) error {
-	if fromPulseNumber < 0 {
-		return errors.New("fromPulseNumber cannot be negative")
-	}
-	if toPulseNumber < 1 {
-		return errors.New("toPulseNumber cannot be less than 1")
-	}
-	if fromPulseNumber > toPulseNumber {
-		return errors.New("fromPulseNumber cannot be greater than toPulseNumber")
+func (e *PlatformExtractor) LoadJetDrops(ctx context.Context, fromPulseNumber int64, toPulseNumber int64) error {
+	e.retrievePulses(ctx, fromPulseNumber, toPulseNumber)
+	return nil
+}
+
+func (e *PlatformExtractor) Stop(ctx context.Context) error {
+	e.cancel()
+	e.startStopMutex.Lock()
```

Comment:
the lock should be before cancel()

Reply:
No, Cancel already have a lock inside, we need to decide do we need this e.hasStarted